include/clap/preset-discovery.h

```diff
@@ -0,0 +1,244 @@
+/**
+   Preset Discovery API.
+
+   Preset Discovery enables a plug-in host to identify where presets are found, what
+   extensions they have, which plug-ins they apply to, and other metadata associated with the
+   presets so that they can be indexed and searched for quickly within the plug-in host's browser.
+
+   This has a number of advantages for the user:
+   - it allows them to browse for presets from one central location in a consistent way
+   - the user can browse for presets without having to commit to a particular plug-in first
+
+   The API works as follow to index presets and presets metadata:
+   1. get the preset discovery factory
+   2. create a preset provider
+   3. list locations and file extensions from the preset provider
+   4. monitor file system changes on the given locations
+   5. read metadata for each preset files
+   6. fetch information about collections from the preset provider
+
+   Then to load a preset, use ext/draft/preset-load.h
+
+   The design of this API deliberately does not define a fixed set tags or categories. It is the
+   plug-in host's job to try to intelligently map the raw list of features that are found for a
+   preset and to process this list to generate something that makes sense for the host's tagging and
+   categorization system. The reason for this is to reduce the work for a plug-in developer to add
+   Preset Discovery support for their existing preset file format and not have to be concerned with
+   all the different hosts and how they want to receive the metadata.
+*/
+
+#pragma once
+
+#include "private/std.h"
+#include "private/macros.h"
+#include "string-sizes.h"
+#include "version.h"
+#include "plugin-features.h"
+#include "preset-features.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** Receiver that receives the metadata for a single preset file.
+ * The host would define the various callbacks in this interface and the preset parser function
+ * would then call them.
+ *
+ * This interface isn't thread-safe.
```

Comment:
What does this imply for the array-getters? e.g. the host might do
```c
uint32_t count = provider->locations_count(provider);
for (uint32_t i = 0; i < count; i +=1) {
    clap_preset_location_t info = {0};
    if (provider->get_location(provider, i, &info)) {
        patchLocationsArray.Add(info);
    }
}
```
Is this guaranteed to get the whole array correctly?

Reply:
Yes it should